### PR TITLE
Moved some APIs around to make `Optional<T>` covariant with respect to T

### PR DESCRIPTION
**Pull request by ashfordneil**
Related Ticket: N/A

Description of Changes:
* Removed .is and .equals and .equals_
* Added .is and .equals to Optionals.ts (note the s)
* Added a test to make sure that optionals are now covariant

Pre-checks:
* [x] Changelog entry added
  * [x] Major version bump of every library we have?
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable): N/A

### CODE REVIEW

**Review by ltrouton — APPROVED** [review]

File: modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts [hunk 1]
```diff
@@ -2,6 +2,22 @@ import * as Arr from './Arr';
 import * as Fun from './Fun';
 import { Optional } from './Optional';
 
+/**
+ **Is** the value stored inside this Optional object equal to `other`?
```
Comment (JamesToohey) [inline modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts:6]: Is `other` referencing something here?

File: modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+interface Cat extends Animal {
+  length: number;
+}
+
+// Sorry for the maths, but really there's a large body of literature working
+// to describe exactly the problem that we're interested in here - and it all
+// comes from maths. There's no advantage in not leaning on that existing work.
+//
+// In particular, we're looking here:
```
Comment (JamesToohey) [inline modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts:18]: I'm not sure this is necessary, the link to the wikipedia article should be fine

**Review by JamesToohey — APPROVED** [review]

File: modules/agar/src/main/ts/ephox/agar/assertions/ApproxStructures.ts [hunk 3]
```diff
@@ -132,7 +132,7 @@ const text = (s: StringAssert, combineSiblings = false): StructAssert => {
       }, (t: string) => {
         let text = t;
         if (combineSiblings) {
-          while (queue.peek().map(SugarNode.isText).is(true)) {
+          while (queue.peek().map(SugarNode.isText).getOr(false)) {
```
Comment (lnewson) [inline modules/agar/src/main/ts/ephox/agar/assertions/ApproxStructures.ts:135]:
```suggestion
          while (queue.peek().exists(SugarNode.isText)) {
```

File: modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts [hunk 4]
```diff
@@ -2,6 +2,22 @@ import * as Arr from './Arr';
 import * as Fun from './Fun';
 import { Optional } from './Optional';
 
+/**
+ **Is** the value stored inside this Optional object equal to `rhs`?
+ */
```
Comment (lnewson) [inline modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts:7]:
```suggestion
/**
 * **Is** the value stored inside this Optional object equal to `rhs`?
 */
```

File: modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts [hunk 5]
```diff
@@ -2,6 +2,22 @@ import * as Arr from './Arr';
...
+export const is = <T>(lhs: Optional<T>, rhs: T, comparator: (a: T, b: T) => boolean = Fun.tripleEquals): boolean =>
+  lhs.exists((left) => comparator(left, rhs));
+
+/**
+ Are these two Optional objects equal? Equality here means either they're both
+ `Some` (and the values are equal under the comparator) or they're both `None`.
+ */
```
Comment (lnewson) [inline modules/katamari/src/main/ts/ephox/katamari/api/Optionals.ts:14]:
```suggestion
/**
 * Are these two Optional objects equal? Equality here means either they're both
 * `Some` (and the values are equal under the comparator) or they're both `None`.
 */
```

File: modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import { assert } from 'chai';
+import { Optional } from 'ephox/katamari/api/Optional';
+import * as Optionals from 'ephox/katamari/api/Optionals';
+
+interface Animal {
+  name: string;
+}
+
+interface Cat extends Animal {
+  length: number;
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts:11]:
```suggestion
  readonly length: number;
```

File: modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { describe, it } from '@ephox/bedrock-client';
+import { assert } from 'chai';
+import { Optional } from 'ephox/katamari/api/Optional';
+import * as Optionals from 'ephox/katamari/api/Optionals';
+
+interface Animal {
+  name: string;
```
Comment (lnewson) [inline modules/katamari/src/test/ts/atomic/api/optional/CovarianceAssertionTest.ts:7]:
```suggestion
  readonly name: string;
```

**Review by lnewson — APPROVED** [review]

**Pull request merged**